### PR TITLE
AD-2 (Implement Authorization)

### PR DESCRIPTION
**Pull request by MinhLu12**
This pull request configures the authentication and authorization for the web API. To get set up, go to the AdminDashboard directory and use dotnet user-secrets to store the hashed password (
1. Navigate to the AdminDashboard project directory
2.  Run  ```dotnet user-secrets init```
3. Run ```dotnet user-secrets "AuthorizationConfiguration:Secret" "<hashed password>"```

Reference to the workflow: https://docs.google.com/document/d/1SUOwBpVA55hXOANGyHlSXm_SZFIHivj3G8LcsxXD0Jk/edit?ts=5dbb400c#heading=h.3kw1g3rjguo9

Let me know what everyone thinks!

### CODE REVIEW

File: AdminDashboard/AdminDashboard.csproj [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
-
   <PropertyGroup>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <NoWarn>$(NoWarn);1591</NoWarn>
   </PropertyGroup>
-
   <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="3.0.0" />
     <PackageReference Include="MySql.Data.EntityFrameworkCore" Version="8.0.18" />
-    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.0.0-rc4" />
```
Comment (MinhLu12) [inline AdminDashboard/AdminDashboard.csproj:14]: For simplicity, I've removed Swagger from this project now that the server-side should be mostly done. It was only there for testing purposes!

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+using System;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace GravitationalTest.BusinessOrchestrators.Users
+{
+    public class LoginOrchestrator : ILoginOrchestrator
+    {
```
Comment (MinhLu12) [inline AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs:15]: Here is the logic that authenticates the user, and if she is authorized, a JWT token will be generated. Some things to note in regards to the problem statement:
1. It is allowed to hard-code the Username (in the future, would probably store it in the database)
2. The password is hashed and stored using the NET standard (pbkdf2)
3. It's okay to hard-code the bearer token in appsettings.json

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                throw new InvalidCredentialsException();
+
+            return GenerateJwtToken();
+        }
+
+        private bool IsPasswordIncorrect(string password)
+        {
+            string savedPasswordHash = Configuration.Secret;
+            byte[] expectedHash = GetBytesFrom(savedPasswordHash);
+            byte[] salt = new byte[16];
```
Comment (alex-kovoy) [inline AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: These magic numbers `20` and `16` appear in a few places. If you are taking shortcuts for simplicity reasons, consider making it easier to follow. Some comments would be helpful

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        private static SecurityTokenDescriptor GenerateTokenDetailsFrom(byte[] key)
+        {
+            return new SecurityTokenDescriptor
+            {
+                Subject = new ClaimsIdentity(new Claim[]
+                {
+                    new Claim(ClaimTypes.Name, Guid.NewGuid().ToString())
+                }),
+                Expires = DateTime.UtcNow.AddDays(7),
+                SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
```
Comment (alex-kovoy) [inline AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs:101]: ` private byte[] GetTokenKey()` returns a bearer token which then is used as a signing key. Hm...why is it called Bearer? Would it be more clear if we call `GetSigningKey`?

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs (hunk 3 above)
Comment (MinhLu12) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: Good catch, would you prefer I use a constant to express the intent or use an inline comment?

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs (hunk 4 above)
Comment (MinhLu12) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs:101]: Yeah that'd make the method much more clear! Changing it.

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs (hunk 3 above)
Comment (alex-kovoy) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: I would do both. Comments provide a better idea about the shortcut.

**Review by alex-kovoy — APPROVED** [review]

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs (hunk 3 above)
Comment (MinhLu12) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: Done!

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            byte[] expectedHash = GetBytesFrom(savedPasswordHash);
+            byte[] salt = new byte[SALT_BYTE_SIZE];
+            Array.Copy(expectedHash, 0, salt, 0, SALT_BYTE_SIZE);
+
+            byte[] actualHash = HashPasswordAttempt(password, salt);
+
+            return CompareHashes(expectedHash, actualHash);
+        }
+
+        private static bool CompareHashes(byte[] expectedHash, byte[] actualHash)
```
Comment (pierrebeaucamp) [inline AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: I dislike the name of this function. Just looking at its signature, it's unclear what `CompareHashes` does - I would expect it returns true if the hashes are equal. Maybe something like `IsEqual` would be better - or just inline the for loop in your `IsPasswordIncorrect` method above

Comment (MinhLu12) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: Yeah, I think I overencapsulated - I put the functionality inline the `IsPasswordIncorrect` function!

**Review by pierrebeaucamp — APPROVED** [review]

**Pull request merged**